### PR TITLE
OSBS-3053: threaded filesystem poll

### PR DESCRIPTION
**Pull request by sosiouxme**
Have DBW use a background thread to track filesystem usage over time so it can be recorded in the build metadata.

### CODE REVIEW

File: atomic_reactor/inner.py [hunk 1]
```diff
@@ -252,6 +254,52 @@ def all_registries(self):
...
+        new_data = dict(
+            mb_free=st.f_bavail * st.f_frsize / mb,
+            mb_total=st.f_blocks * st.f_frsize / mb,
+            mb_used=(st.f_blocks - st.f_bfree) * st.f_frsize / mb,
+            inodes_free=st.f_ffree,
+            inodes_total=st.f_files,
+            inodes_used=st.f_files - st.f_ffree,
+        )
+        for key in ["mb_total", "mb_used", "inodes_total", "inodes_used"]:
+            if new_data[key] > data.get(key):
```
Comment (lcarva) [inline atomic_reactor/inner.py]: `dict.get` returns `None` if key is set. In python2 `None` is evaluated as greater than for any integer. In python 3 it raises `TypeError`. Does it make sense to use `0` as the default value (`data.get(key, 0)`)? Or is checking for falsy value better?

Comment (sosiouxme) [reply atomic_reactor/inner.py]: Ah, did not know that, thanks. I think `data.get(key, 0)` is better, can use it the same here and below.

Comment (sosiouxme) [reply atomic_reactor/inner.py]: Do PR tests run under both python 2 and 3?

Comment (lcarva) [reply atomic_reactor/inner.py]: I think if you use it below, it's going to mess up the "min" check.
I'm wondering if you need `sys.maxsize` for that one.
*Note that this is max **int** size, you can still get a larger value but it'll be of type **long**.

Comment (lcarva) [reply atomic_reactor/inner.py]: Yes, they run for both 2 and 3. It looks like the tests for python 3 failed: https://travis-ci.org/projectatomic/atomic-reactor/builds/367111697?utm_source=github_status&utm_medium=notification

File: atomic_reactor/inner.py [hunk 2]
```diff
@@ -252,6 +254,52 @@ def all_registries(self):
...
+class FSWatcher(threading.Thread):
+
+    def __init__(self, *args, **kwargs):
+        super(FSWatcher, self).__init__(*args, **kwargs)
+        self.daemon = True  # exits whenever the process exits
+        self._lock = threading.RLock()
+        self._data = {}
+
+    def run(self):
+        while True:
```
[the remark this answers is not in the document]
Comment (sosiouxme) [reply atomic_reactor/inner.py]: Ah, I hadn't thought about dangling threads in testing (overhead would still be minimal, but it does feel untidy). The API doesn't provide an obvious way to outright kill a thread, but I guess it would be easy enough to have it exit.

**Comment by sosiouxme** [comment]
Updated and hopefully complete -- please review.

File: atomic_reactor/inner.py [hunk 3]
```diff
@@ -252,6 +255,62 @@ def all_registries(self):
...
+    def __init__(self, *args, **kwargs):
+        super(FSWatcher, self).__init__(*args, **kwargs)
+        self.daemon = True  # exits whenever the process exits
+        self._lock = threading.Lock()
+        self._done = False
+        self._data = {}
+
+    def run(self):
+        """ Overrides parent method to implement thread's functionality. """
+        while True:
```
Comment (lcarva) [inline atomic_reactor/inner.py]: `while not self._done` ?

Comment (sosiouxme) [reply atomic_reactor/inner.py]: I guess that's prettier :+1:

File: atomic_reactor/inner.py [hunk 4]
```diff
@@ -452,6 +513,7 @@ def build_docker_image(self):
             finally:
                 self.source.remove_tmpdir()
 
+            self.fs_watcher.finish()
```
Comment (lcarva) [inline atomic_reactor/inner.py]: Should this go under the `finally` block above? If an exit plugin fails, this line is not reached.

File: atomic_reactor/plugins/exit_store_metadata_in_osv3.py [hunk 5]
```diff
@@ -223,7 +233,8 @@ def run(self):
             "base-image-name": base_image_name,
             "image-id": self.workflow.builder.image_id or '',
             "digests": json.dumps(self.get_pullspecs(self.get_digests())),
-            "plugins-metadata": json.dumps(self.get_plugin_metadata())
+            "plugins-metadata": json.dumps(self.get_plugin_metadata()),
+            "filesystem": json.dumps(self.get_filesystem_metadata()),
```
Comment (lcarva) [inline atomic_reactor/plugins/exit_store_metadata_in_osv3.py]: It would be great to add a unit test for ensuring this annotation is set properly.

File: requirements.txt [hunk 6]
```diff
@@ -4,3 +4,4 @@ dockerfile-parse>=0.0.5
 jsonschema
 PyYAML
 six
+backports.ssl-match-hostname==3.4.0.2
```
Comment (lcarva) [inline requirements.txt]: Why is this needed?

Comment (sosiouxme) [reply requirements.txt]: It's not, I'm a doofus

File: atomic_reactor/inner.py (hunk 4 above)
Comment (sosiouxme) [reply atomic_reactor/inner.py]: That would be unlikely to cause much concern :) but sure, to be tidy.

**Comment by sosiouxme** [comment]
Thanks, I think I've addressed everything again.

**Comment by lcarva** [comment]
Updated release nodes draft:
> - New Build annotation `filesystem` will provide consumed storage stats for each build
